oiosaml/src/main/java/dk/gov/oio/saml/session/database/DatabaseSessionHandler.java

```diff
@@ -0,0 +1,471 @@
+ * created by Trifork A/S are Copyright (C) 2009 Danish National IT
+ * and Telecom Agency (http://www.itst.dk). All Rights Reserved.
+ *
+ * Contributor(s):
+ *   Joakim Recht <jre@trifork.com>
+ *   Rolf Njor Jensen <rolf@trifork.com>
+ *
+ */
+package dk.gov.oio.saml.session.database;
+
+import dk.gov.oio.saml.audit.AuditService;
+import dk.gov.oio.saml.model.NSISLevel;
+import dk.gov.oio.saml.service.OIOSAML3Service;
+import dk.gov.oio.saml.session.AssertionWrapper;
+import dk.gov.oio.saml.session.AuthnRequestWrapper;
+import dk.gov.oio.saml.session.LogoutRequestWrapper;
+import dk.gov.oio.saml.session.SessionHandler;
+import dk.gov.oio.saml.session.inmemory.TimeOutWrapper;
+import dk.gov.oio.saml.util.InternalException;
+import dk.gov.oio.saml.util.StringUtil;
+import net.shibboleth.utilities.java.support.xml.SerializeSupport;
+import net.shibboleth.utilities.java.support.xml.XMLParserException;
+import org.apache.commons.codec.binary.Base64;
+import org.opensaml.core.xml.config.XMLObjectProviderRegistrySupport;
+import org.opensaml.core.xml.io.MarshallingException;
+import org.opensaml.core.xml.io.UnmarshallingException;
+import org.opensaml.core.xml.util.XMLObjectSupport;
+import org.opensaml.saml.saml2.core.Assertion;
+import org.opensaml.saml.saml2.core.AuthnRequest;
+import org.opensaml.saml.saml2.core.LogoutRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.http.HttpSession;
+import javax.sql.DataSource;
+import java.io.ByteArrayInputStream;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.sql.*;
+import java.time.Clock;
+import java.util.Date;
+
+public class DatabaseSessionHandler implements SessionHandler {
+    private static final Logger log = LoggerFactory.getLogger(DatabaseSessionHandler.class);
+
+    private final DataSource ds;
+
+    public DatabaseSessionHandler(DataSource ds) {
+        log.debug("Created database session handler");
+        this.ds = ds;
+    }
+    /**
+     * Set AuthnRequest on the current session
+     *
+     * @param session HTTP session
+     * @param request {@link AuthnRequestWrapper}
+     * @throws InternalException on failure to persist request
+     */
+    @Override
+    public void storeAuthnRequest(HttpSession session, AuthnRequestWrapper request) throws InternalException {
+        if (null == request || null == request.getId()) {
+            log.warn("Ignore AuthRequest with null value or missing ID");
+            return;
+        }
+        try (Connection connection=ds.getConnection()){
+            connection.setAutoCommit(true);
+
+            AuthnRequestWrapper authnRequest = getAuthnRequest(session);
+            if (null != authnRequest) {
+                log.debug("AuthRequest '{}' will replace '{}'", request.getId(), authnRequest.getId());
+                try (PreparedStatement ps = connection.prepareStatement("DELETE FROM authn_requests_tbl WHERE session_id = ?")) {
```

Comment:
Don't know where this is used but errors in these SQL queries are discovered at runtime and puts restrictions on things like table naming

Reply:
No better or worse than the old implementation and if the service provider wants a more flexibility option then they are free to create new SessionHandler implementation either as an extension or borrow our code implement their own.

Reply:
Since it's JDBC we are stuck, but this should have been done with JPA, maybe hibernate.

Reply:
I completely agree, just had to mention it